### PR TITLE
Add geothermal-sourced central heat pumps

### PR DESCRIPTION
**Pull request by amos-schledorn**
This feature adds geothermal-sourced central heating based on utilisation potentials by [Manz et al.](https://www.sciencedirect.com/science/article/pii/S0960148124001769) and technology data from the DEA techn. catalogue.

## Changes proposed in this Pull Request
-`retrieve_fraunhofer_heat_source_utilisation_potentials`: retrieve heat source potentials for those heat sources set in `config` under `sector:district_heating:fraunhofer_heat_utilisation_potentials` if the respective source is listed in the `config` under `sector:heat_pump_sources:urban_central`
-`build_heat_source_potentials`: map utilisation potentials to onshore-regions and scale them by a) FLH and b) unit (input is TWh)
- `build_cop_profiles`: a) COP for fixed-source temperature geothermal heat is computed as for air-sourced heat pumps, b) for all sources: COP is set to zero if source temperature exceeds forward temperature
- `build_direct_heat_source_utilisation_potentials`: generates profiles in the same format as COP profiles. 1, when (where) forward temperature above source temperature, 0 otherwise.
- `prepare_sector_network`: If heat pump source is in `fraunhofer_heat_sources`, then add a generator. The corresponding heat pumps convert `1 MWe` (electricity) + `(COP - 1) MWth` (geothermal heat) to `COP MWth` usable central heat. Also add a direct-utilisation-link (using the aforementioned profiles) that converts the heat source to urban central heat 1-to-1.

## Notes
- DEA technology data refers to 1200m deep geothermal district heating (in contrast to 2000-3000m assumed in Manz et al.)
- The results for low spatial resolution can be inaccurate: Geothermal heat is overused as relatively large areas are modelled as heat lakes. This could be remedied limiting the geothermal potential pre-clustering to the NUTS3 heat demand (and then summing the limited geothermal wattage). That's why I would suggest turning geothermal heat off by default.
- Any heat source from the Manz et al. (2024) can be used by adjusting the config accordingly.
- This feature necessitates an additional technology for direct geothermal heat utilisation (c.f. [feature in technology-data](https://github.com/PyPSA/technology-data/pull/153), which is why the related `config` setting points to the `technology-data` master, rather than the latest release. I'm not entirely sure how to resolve that. 

## Checklist

- [x] I tested my contribution locally and it works as intended.
- [x] Code and workflow changes are sufficiently documented.
- [x] Changed dependencies are added to `envs/environment.yaml`.
- [x] Changes in configuration options are added in `config/config.default.yaml`.
- [x] Changes in configuration options are documented in `doc/configtables/*.csv`.
- [x] Sources of newly added data are documented in `doc/data_sources.rst`.
- [x] A release note `doc/release_notes.rst` is added.

## Testing
I've used the following config for testing:
```run:
  name: 
    - no-geothermal-365h
    - geothermal-365h
  scenarios:
    enable: true
    file: config/scenarios.yaml

scenario:
  clusters:
    - 6
    - 20
    - 60
  planning_horizons:
    - 2030

foresight: overnight
countries: ['DE', 'DK']

clustering:
  temporal:
    resolution_sector: 365h
```

with the following `scenarios.yaml`:
```no-geothermal-365h:
  sector:
    heat_pump_sources:
      urban central:
      - air

geothermal-365h:
  sector:
    heat_pump_sources:
      urban central:
      - air
      - geothermal
```

Results for the `no-geothermal-365h` scenario seem like a reasonable baseline:
image

In the `geothermal-365h` scenario, it becomes apparent that we need to use a high spatial resolution, since geothermal heat utilisation varies strongly going from 6 to 20 to 60 clusters (note that the order in the plot is 20, 6, 60):
image

I've quickly checked this and we can see that, if for each region, the geothermal potential is capped by the heat demand (i.e. we take the max. of the two per region) and sum over regions, this number decreases with an increasing number of clusters:
image

Hence, I propose to turn this feature off by default (I'm just leaving this turned on for the CI but would change this before merging - **so please do not merge (or turn off geothermal heat in the config)**.

### CODE REVIEW

**Review by cpschau — CHANGES_REQUESTED** [review]
Looks very good and clean to me! I have only two formal remarks about the data documentation and the technology data version referenced in the config. I would agree to to keep it turned off, until a reasonable value is found for the upper bound. For the latter, one could maybe use the [open data from the seenergies project](https://s-eenergies-open-data-euf.hub.arcgis.com/datasets/b62b8ad79f0e4ae38f032ad6aadb91a0_0/about).

File: config/config.default.yaml [hunk 1]
```diff
@@ -827,7 +840,7 @@ industry:
 # docs in https://pypsa-eur.readthedocs.io/en/latest/configuration.html#costs
 costs:
   year: 2030
-  version: v0.9.2
+  version: master
```
[the remark this answers is not in the document]
Comment (lkstrp) [reply config/config.default.yaml:843]:
```suggestion
  version: v0.9.3
```
Done

File: doc/data-retrieval.rst [hunk 2]
```diff
@@ -9,12 +9,21 @@ Specific retrieval rules
...
+- **Description:** Europe's NUTS administrative regions.
+
+
 ``data/nuts``
 
 - **Source:** GISCO
 - **Link:** https://gisco-services.ec.europa.eu/distribution/v2/nuts/download/
-- **License:** `custom <https://ec.europa.eu/eurostat/web/gisco/geodata/reference-data/administrative-units-statistical-units>`__
-- **Description:** Europe's NUTS administrative regions.
+- **License:** `CC BY 4.0 <https://ec.europa.eu/eurostat/web/gisco/geodata/reference-data/administrative-units-statistical-unitshttps://creativecommons.org/licenses/by/4.0/>`__
```
[the remark this answers is not in the document]
Comment (amos-schledorn) [reply doc/data-retrieval.rst]: Ooops, thanks for spotting this. Should be fixed now.

**Comment by amos-schledorn** [comment]
> Looks very good and clean to me! I have only two formal remarks about the data documentation and the technology data version referenced in the config. I would agree to to keep it turned off, until a reasonable value is found for the upper bound. For the latter, one could maybe use the [open data from the seenergies project](https://s-eenergies-open-data-euf.hub.arcgis.com/datasets/b62b8ad79f0e4ae38f032ad6aadb91a0_0/about).

Thanks for the review, @cpschau! I've fixed the docs and `technology-data` version and will merge, after another test run.
I agree regarding bounds on heat source usage and opened a separate issue: #1443

File: rules/build_sector.smk [hunk 3]
```diff
@@ -284,6 +284,29 @@ rule build_central_heating_temperature_profiles:
         "../scripts/build_central_heating_temperature_profiles/run.py"
 
 
+rule build_heat_source_potentials:
+    params:
+        fraunhofer_heat_sources=config_provider(
+            "sector", "district_heating", "fraunhofer_heat_utilisation_potentials"
+        ),
+        heat_source="{heat_source}",
```
Comment (fneum) [inline rules/build_sector.smk:292]: Why is that necessary? Could it not just be accessed in the script with `snakemake.wildcards.heat_source`? Just curious -- fine to leave as is if deemed necessary.

File: scripts/build_heat_source_potentials/run.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        scaling_factor=get_unit_conversion_factor(
+            input_unit=snakemake.params.fraunhofer_heat_sources[
+                snakemake.params.heat_source
+            ]["unit"],
+            output_unit="MWh",
+        )
+        / snakemake.params.fraunhofer_heat_sources[snakemake.params.heat_source][
+            "full_load_hours"
+        ],
+    ).data_in_regions_scaled
```
Comment (fneum) [inline scripts/build_heat_source_potentials/run.py:89]: More a style remark, but this could be a little less nested to improve readability.

File: config/config.default.yaml [hunk 5]
```diff
@@ -480,6 +480,18 @@ sector:
...
+      geothermal:
+        # activate for 85C hydrothermal
+        # key: hydrothermal_85
+        # constant_temperature_celsius: 85
+        key: hydrothermal_65
+        constant_temperature_celsius: 65
+        column_name: Energy_TWh
+        unit: TWh
+        full_load_hours: 4000
+    direct_utilisation_heat_sources:
```
Comment (fneum) [inline config/config.default.yaml:493]: This setting is not documented in `sector.csv` is it?

File: rules/build_sector.smk (hunk 3 above)
Comment (amos-schledorn) [reply rules/build_sector.smk:292]: Done - didn't know I could do that:)

**Review by fneum — APPROVED** [review]
